### PR TITLE
checkAars task

### PR DESCRIPTION
**Pull request by tek**
collect all AarLibrary deps and determines duplicates with different versions

searches the complete sbt project dep tree for the source of the duplicate depspecs

### CODE REVIEW

File: src/dependencies.scala [hunk 1]
```diff
@@ -150,5 +173,41 @@ object Dependencies {
...
+      partsCover && (
+        parts.length == otherParts.length ||
+        (parts.length < otherParts.length) && parts.lastOption.exists(_ == "+")
+      )
+    }
+
+    def covers(other: ModuleID) = {
+      id.organization == other.organization && id.name == other.name &&
+        revCovers(other.revision)
+    }
```
Comment (pfn) [inline src/dependencies.scala]: I don't care too much for the revCovers and covers naming. something like "revMatches" and "matches" would be preferable

File: src/dependencies.scala [hunk 2]
```diff
@@ -150,5 +173,41 @@ object Dependencies {
...
+
+  implicit class ProjectRefOps(project: ProjectRef)
+  (implicit struct: sbt.BuildStructure)
+  {
+    def resolved = Project.getProject(project, struct)
+
+    def deps = resolved map(_.dependencies) getOrElse(Nil) map(_.project)
+
+    def deepDeps: Seq[ProjectRef] =
+      ((deps map(_.deepDeps) flatten) :+ project).distinct
```
Comment (pfn) [inline src/dependencies.scala]: deps flatMap (_.deepDeps)

File: src/dependencies.scala [hunk 3]
```diff
@@ -150,5 +173,41 @@ object Dependencies {
...
+
+    def dependsOnAar(aar: AarLibrary) = {
+      (sbt.Keys.libraryDependencies in project)
+        .get(struct.data)
+        .exists(_.find(_.covers(aar.moduleID)).isDefined)
```
Comment (pfn) [inline src/dependencies.scala]: .exists(_.exists(_.covers(aar.moduleID)))

File: src/rules.scala [hunk 4]
```diff
@@ -430,6 +434,7 @@ object Plugin extends sbt.Plugin {
     libraryProjects          := localProjects.value ++ apklibs.value ++ aars.value,
     libraryProject          <<= properties { p =>
       Option(p.getProperty("android.library")) exists { _.equals("true") } },
+    checkAars               <<= checkAarsTaskDef,
```
Comment (pfn) [inline src/rules.scala]: hook checkAars into either `aars` or `collectResources`, whichever makes sense for its dependencies

File: src/tasks.scala [hunk 5]
```diff
@@ -261,6 +261,48 @@ object Tasks {
...
+
+  val checkAarsTaskDef = Def.task {
+    implicit val log = streams.value.log
+    implicit val struct = buildStructure.value
+    val deps = thisProjectRef.value.deepDeps
+    def revDependents(aar: AarLibrary) = deps.filter(_.dependsOnAar(aar))
+    def pkgDependents(data: (String, Seq[AarLibrary])) = {
+      (data._1, data._2, data._2 map(revDependents))
+    }
+    val broken = duplicateAars.value
```
Comment (pfn) [inline src/tasks.scala]: neat, didn't know you could `.value` on arbitrary task values

Comment (tek) [inline src/tasks.scala]: Alright, I'm on vacation until Wednesday, will fix it then

**Pull request closed without merging**